libs/storage/src/ipfs/ipfs.cluster.service.ts

```diff
@@ -0,0 +1,544 @@
+      cidBytes: result.data?.cidBytes || Buffer.from([]),
+      fileName: filename,
+      size: result.data?.size || fileBuffer.length,
+      hash: '',
+    };
+  }
+
+  /**
+   * Check if a CID is pinned in the cluster
+   */
+  public async isPinned(cid: string): Promise<boolean> {
+    try {
+      const pinStatus = await this.checkPinStatus(cid);
+      return this.isPinnedInCluster(pinStatus.data);
+    } catch (err: any) {
+      // If CID is not found, it's not pinned
+      if (err.message?.includes('not found') || err.message?.includes('404')) {
+        return false;
+      }
+      throw err;
+    }
+  }
+
+  /**
+   * Check if any peers in the peer_map have status 'pinned' or 'pinning'
+   */
+  private isPinnedInCluster(clusterData: any): boolean {
+    if (!clusterData || !clusterData.peer_map) {
+      return false;
+    }
+
+    const peerMap = clusterData.peer_map;
+
+    // Iterate through all peer entries in the peer_map
+    for (const peerId in peerMap) {
+      const peerInfo = peerMap[peerId];
+      if (peerInfo && (peerInfo.status === 'pinned' || peerInfo.status === 'pinning')) {
+        this.logger.debug(`Found pinned peer: ${peerId} with status: ${peerInfo.status}`);
+        return true;
+      }
+    }
+
+    this.logger.debug('No pinned peers found in peer_map');
+    return false;
+  }
+
+  private async checkPinStatus(cid: string): Promise<any> {
+    const url = `${this.config.ipfsEndpoint}/pins/${cid}`;
+    return this.makeRequestWithRetry(url, { method: 'GET' });
+  }
+
+  /**
+   * Execute request with retry logic and exponential backoff
+   */
+  private async makeRequestWithRetry(
+    url: string,
+    options: RequestInit = {},
+  ): Promise<{
+    status: number;
+    statusText: string;
+    headers: Record<string, string>;
+    data: any;
+    error?: string;
+    stack?: string;
+  }> {
+    let lastError: Error | null = null;
+
+    for (let attempt = 0; attempt <= this.config.retryAttempts; attempt++) {
+      try {
+        this.logger.debug(`Attempt ${attempt + 1}/${this.config.retryAttempts + 1} for ${url}`);
+        return await this.makeRequest(url, options);
+      } catch (error) {
+        lastError = error as Error;
+        const errorMessage = error instanceof Error ? error.message : String(error);
+        
+        // Don't retry on the last attempt
+        if (attempt < this.config.retryAttempts) {
+          const delay = Math.min(Math.pow(2, attempt) * 100, 5000); // Cap at 5 seconds
+          this.logger.debug(`Request failed, retrying in ${delay}ms: ${errorMessage}`);
+          await this.delay(delay);
+        } else {
+          this.logger.error(`All ${this.config.retryAttempts + 1} attempts failed for ${url}: ${errorMessage}`);
+        }
+      }
+    }
+
+    throw lastError;
+  }
+
+  /**
+   * Delay utility for retry backoff
+   */
+  private delay(ms: number): Promise<void> {
+    return new Promise(resolve => setTimeout(resolve, ms));
+  }
```

Comment:
\<rant>You'd think they would have provided an async `delayMs` function or something in Node.js by now instead of always having to do this garbage\</rant>

nit: this is probably better off as a file-scope function instead of a class private. Or better yet, in our utils library?